core/test/attributes_test.cpp

```diff
@@ -0,0 +1,121 @@
+
+using namespace scipp;
+using namespace scipp::core;
+
+class AttributesTest : public ::testing::Test {
+protected:
+  const Variable scalar = makeVariable<double>(1);
+  const Variable varX = makeVariable<double>({Dim::X, 2}, {2, 3});
+  const Variable varZX =
+      makeVariable<double>({{Dim::Y, 2}, {Dim::X, 2}}, {4, 5, 6, 7});
+};
+
+TEST_F(AttributesTest, dataset_attrs) {
+  Dataset d;
+  d.setAttr("scalar", scalar);
+  d.setAttr("x", varX);
+  ASSERT_EQ(d.attrs().size(), 2);
+  ASSERT_TRUE(d.attrs().contains("scalar"));
+  ASSERT_TRUE(d.attrs().contains("x"));
+  ASSERT_EQ(d.dimensions(),
+            (std::unordered_map<Dim, scipp::index>{{Dim::X, 2}}));
+  d.eraseAttr("scalar");
+  d.eraseAttr("x");
+  ASSERT_EQ(d.attrs().size(), 0);
+  ASSERT_EQ(d.dimensions(), (std::unordered_map<Dim, scipp::index>{}));
+}
+
+TEST_F(AttributesTest, dataset_item_attrs) {
+  Dataset d;
+  d.setData("a", varX);
+  d["a"].attrs().set("scalar", scalar);
+  d["a"].attrs().set("x", varX);
+  d.attrs().set("dataset_attr", scalar);
+
+  ASSERT_FALSE(d.attrs().contains("scalar"));
+  ASSERT_FALSE(d.attrs().contains("x"));
+
+  ASSERT_EQ(d["a"].attrs().size(), 2);
+  ASSERT_TRUE(d["a"].attrs().contains("scalar"));
+  ASSERT_TRUE(d["a"].attrs().contains("x"));
+  ASSERT_FALSE(d["a"].attrs().contains("dataset_attr"));
+
+  d["a"].attrs().erase("scalar");
+  d["a"].attrs().erase("x");
+  ASSERT_EQ(d["a"].attrs().size(), 0);
+}
+
+TEST_F(AttributesTest, dataset_item_attrs_dimensions_exceeding_data) {
+  Dataset d;
+  d.setData("scalar", scalar);
+  EXPECT_THROW(d["scalar"].attrs().set("x", varX), except::DimensionError);
+}
+
+TEST_F(AttributesTest, slice_dataset_item_attrs) {
+  Dataset d;
+  d.setData("a", varZX);
+  d["a"].attrs().set("scalar", scalar);
+  d["a"].attrs().set("x", varX);
+
+  // Same behavior as coord slicing:
+  // - Lower-dimensional attrs are not hidden by slicing.
+  // - Non-range slice hides attribute.
+  // The alternative would be to handle attributes like data, but at least for
+  // now coord-like handling appears to make more sense.
+  ASSERT_TRUE(d["a"].slice({Dim::X, 0}).attrs().contains("scalar"));
+  ASSERT_FALSE(d["a"].slice({Dim::X, 0}).attrs().contains("x"));
+  ASSERT_TRUE(d["a"].slice({Dim::X, 0, 1}).attrs().contains("scalar"));
+  ASSERT_TRUE(d["a"].slice({Dim::X, 0, 1}).attrs().contains("x"));
+  ASSERT_TRUE(d["a"].slice({Dim::Y, 0}).attrs().contains("scalar"));
+  ASSERT_TRUE(d["a"].slice({Dim::Y, 0}).attrs().contains("x"));
+  ASSERT_TRUE(d["a"].slice({Dim::Y, 0, 1}).attrs().contains("scalar"));
+  ASSERT_TRUE(d["a"].slice({Dim::Y, 0, 1}).attrs().contains("x"));
+}
+
+TEST_F(AttributesTest, binary_ops_in_place) {
+  Dataset d1;
+  d1.setData("a", varX);
+  d1["a"].attrs().set("a_attr", scalar);
+  d1.attrs().set("dataset_attr", scalar);
+
+  Dataset d2;
+  d2.setData("a", varX);
+  d2["a"].attrs().set("a_attr", varX);
+  d2.attrs().set("dataset_attr", varX);
+
+  auto result(d1);
+  result += d2;
```

Comment:
Maybe worth checking here some intermediate results? The labels are just compared right? But what are the attributes of `resut` at this point here? I guess it is not replaced if it already exists? So you could also check here that
```C++
  EXPECT_EQ(result.attrs()["dataset_attr"], scalar);
  EXPECT_EQ(result["a"].attrs()["a_attr"], scalar);
```
In addition, what happens if you have
```C++
  Dataset d1;
  d1.setData("a", varX);
  d1["a"].attrs().set("a_attr1", scalar);
  d1.attrs().set("dataset_attr1", scalar);

  Dataset d2;
  d2.setData("a", varX);
  d2["a"].attrs().set("a_attr2", varX);
  d2.attrs().set("dataset_attr2", varX);

  auto result(d1);
  result += d2;
```
at this point, does `result` hold both `"dataset_attr1"` and `"dataset_attr2"`?

Reply:
Not sure what is gained by checking intermediate results, unless you expect one operator to magically restore things that a previous one removed?

Attributes from `d2` are always ignored, regardless of the name. I will clarfiy this in the test.

Reply:
Ok, making the test a bit clearer would work